File: app/src/lib/taxCalc.ts
/**
 * Israeli equity-compensation tax math. Mirrors `tax_calc.py` in
 * `.cursor/skills/jfrog-equity-analyzer/`. Constants and rules updated for
 * 2026 tax year — change one place, change both.
 */

export const SURTAX_THRESHOLD_NIS = 721_560;
export const CAPITAL_GAINS_RATE_BASE = 0.25;
export const CAPITAL_GAINS_RATE_HIGH = 0.30;

// Bituach Leumi + Mas Briut (employee share) — 2026 rates
export const BL_REDUCED_CEILING = 7_703;
export const BL_MONTHLY_CEILING = 51_910;
export const BL_HEALTH_REDUCED_TOTAL = 0.0427; // BL 1.04% + health 3.23%
export const BL_HEALTH_FULL_TOTAL = 0.1217;    // BL 7.00% + health 5.17%

/**
 * Combined BL + health tax on a single month's income (employee share, 2026).
 *
 * Reduced rates (4.27%) apply to the first 7,703 ₪, full rates (12.17%)
 * from 7,703 ₪ up to the monthly ceiling of 51,910 ₪. No contribution above
 * the ceiling.
 */
export function bituachLeumiEmployee(monthlyIncomeNIS: number): number {
  if (monthlyIncomeNIS <= 0) return 0;
  const reduced = Math.min(monthlyIncomeNIS, BL_REDUCED_CEILING);
  const aboveReduced = Math.max(
    0,
    Math.min(monthlyIncomeNIS, BL_MONTHLY_CEILING) - BL_REDUCED_CEILING,
  );
  return reduced * BL_HEALTH_REDUCED_TOTAL + aboveReduced * BL_HEALTH_FULL_TOTAL;
}

// 2026 piecewise income-tax brackets (income tax only, before BL/health)
const INCOME_TAX_BRACKETS_2026: Array<[number, number]> = [
  [84_120, 0.10],
  [120_720, 0.14],
  [228_000, 0.20],
  [301_200, 0.31],
  [560_280, 0.35],
  [721_560, 0.47],
  [Infinity, 0.50], // 47% + 3% yasaf at the top
];

/**
 * Marginal income-tax rate (NOT including BL/health) at a given annual gross.
 * Returns the rate of the bracket in which the last NIS of income falls.
 */
export function incomeTaxMarginalRate(annualSalaryNIS: number): number {
  for (const [top, rate] of INCOME_TAX_BRACKETS_2026) {
    if (annualSalaryNIS <= top) return rate;
  }
  return 0.50;
}

/**
 * Combined marginal rate for פירותי (ordinary-income) equity: income tax +
 * BL/health applicable to equity income.
 *
 * BL/health ceiling rule: if the employee's regular monthly salary
 * (annualSalaryNIS / 12) already equals or exceeds the monthly ceiling
 * (51,910 ₪ in 2026), the RSU/option/ESPP ordinary-income portion carries
 * zero additional BL/health. Only income tax applies.
 *
 * For salaries below the ceiling, BL/health at the full rate (12.17%) is
 * added — all JFrog employees earn well above the reduced-rate threshold
 * (7,703 ₪/month) so only the full rate is relevant.
 */
export function marginalRate(annualSalaryNIS: number): number {
  const itRate = incomeTaxMarginalRate(annualSalaryNIS);
  const monthlySalary = annualSalaryNIS / 12;
  const blRate = monthlySalary >= BL_MONTHLY_CEILING ? 0 : BL_HEALTH_FULL_TOTAL;
  return itRate + blRate;
}

/**
 * Effective capital-gains rate for a Section 102 honi sale.
 * 25% by default, 30% only for controlling shareholders (10%+ holdings, ever).
 * Portfolio-level מס יסף (3% + 2% since 2025) is computed separately via `computeSurtax`.
 */
export function capitalGainsRate(
  _annualSalaryNIS: number,
  controllingShareholder = false,
): number {
  if (controllingShareholder) return CAPITAL_GAINS_RATE_HIGH;
  return CAPITAL_GAINS_RATE_BASE;
}

/** Inputs for portfolio-level יסף per section 121ב (הוראת ביצוע 5/2025). */
export interface SurtaxInput {
  salaryNIS: number;
  /** Gross taxable proceeds from simulated JFrog sales — all tracks (employment + capital), NIS. */
  jfrogSaleTotalIncomeNIS: number;
  /** Capital-source gross from JFrog sales only (capital-track 102 הוני RSU, options spread, ESPP הוני), NIS. */
  jfrogSaleCapitalSourceNIS: number;
  otherCapitalIncomeNIS?: number;
}

export interface SurtaxResult {
  totalIncomeNIS: number;
  capitalIncomeTotalNIS: number;
  yasaf3NIS: number;
  yasaf2NIS: number;
  totalNIS: number;
}

const SURTAX_GENERAL_RATE = 0.03;
const SURTAX_CAPITAL_EXTRA_RATE = 0.02;

export function computeSurtax(input: SurtaxInput): SurtaxResult {
  const other = input.otherCapitalIncomeNIS ?? 0;
  const totalIncomeNIS = input.salaryNIS + input.jfrogSaleTotalIncomeNIS + other;
  const capitalIncomeTotalNIS = input.jfrogSaleCapitalSourceNIS + other;
  const yasaf3NIS = SURTAX_GENERAL_RATE * Math.max(0, totalIncomeNIS - SURTAX_THRESHOLD_NIS);
  const yasaf2NIS = SURTAX_CAPITAL_EXTRA_RATE * Math.max(0, capitalIncomeTotalNIS - SURTAX_THRESHOLD_NIS);
  const totalNIS = yasaf3NIS + yasaf2NIS;
  return { totalIncomeNIS, capitalIncomeTotalNIS, yasaf3NIS, yasaf2NIS, totalNIS };
}

/** Per-lot gross buckets for יסף: total (3% base) vs capital-source (2% base, ex salary). */
export interface LotIncomeBreakdown {
  totalIncomeNIS: number;
  capitalIncomeNIS: number;
}

export function rsuLotIncome(
  _fmvAtGrant: number,
  salePrice: number,
  qty: number,
  fx: number,
  capitalTrack: boolean,
): LotIncomeBreakdown {
  const totalIncomeNIS = salePrice * qty * fx;
  return {
    totalIncomeNIS,
    capitalIncomeNIS: capitalTrack ? totalIncomeNIS : 0,
  };
}

export function optionLotIncome(
  strike: number,
  salePrice: number,
  qty: number,
  fx: number,
): LotIncomeBreakdown {
  const spread = Math.max(0, salePrice - strike);
  const v = spread * qty * fx;
  return { totalIncomeNIS: v, capitalIncomeNIS: v };
}

export function esppLotIncome(
  _purchaseDateFmv: number,
  salePrice: number,
  qty: number,
  fx: number,
  capitalTrack: boolean,
): LotIncomeBreakdown {
  const totalIncomeNIS = salePrice * qty * fx;
  return {
    totalIncomeNIS,
    capitalIncomeNIS: capitalTrack ? totalIncomeNIS : 0,
  };
}

/**
 * Section 102 capital track requires 24 months from the END OF THE TAX YEAR
 * of grant — not 24 calendar months from the grant date. A grant on
 * 2024-03-15 qualifies for capital treatment from 2026-12-31.
 */
export function isCapitalTrack(grantDate: Date, today: Date = new Date()): boolean {
  const qualifying = new Date(grantDate.getFullYear() + 2, 11, 31);
  return today.getTime() >= qualifying.getTime();
}

/**
 * Net per share for an RSU grant.
 *
 * Capital track + price ≥ FMV-at-grant: FMV portion taxed at marginal,
 * appreciation taxed at the capital-gains rate.
 *
 * Capital track + price < FMV-at-grant (sell-below-FMV cap): all proceeds
 * taxed at marginal, capital component clamps to zero — no synthetic
 * capital loss is recognized inside Section 102.
 *
 * Ordinary track: full sale price taxed at marginal.
 */
export function rsuNetPerShare(
  fmvAtGrant: number,
  salePrice: number,
  rate: number,
  cgRate: number,
  capitalTrack: boolean,
): number {
  if (!capitalTrack) return salePrice * (1 - rate);
  if (salePrice < fmvAtGrant) return salePrice * (1 - rate);
  const empTax = fmvAtGrant * rate;
  const cgTax = (salePrice - fmvAtGrant) * cgRate;
  return Math.max(0, salePrice - empTax - cgTax);
}

export function rsuEffectiveTaxRate(
  fmvAtGrant: number,
  salePrice: number,
  rate: number,
  cgRate: number,
  capitalTrack: boolean,
): number {
  if (salePrice <= 0) return 0;
  const net = rsuNetPerShare(fmvAtGrant, salePrice, rate, cgRate, capitalTrack);
  return 1 - net / salePrice;
}

/**
 * Net per share for an NQ option (assumes pre-IPO grant where strike ≈ FMV at
 * grant, so the employment-income component is ~0 and the entire spread is
 * capital gain).
 */
export function optionNetPerShare(strike: number, salePrice: number, cgRate: number): number {
  if (salePrice <= strike) return 0;
  return (salePrice - strike) * (1 - cgRate);
}

export function optionEffectiveTaxRate(strike: number, salePrice: number, cgRate: number): number {
  const spread = salePrice - strike;
  if (spread <= 0) return 0;
  // Tax is cgRate% of the spread only — the strike is the cost basis paid out-of-pocket.
  // Effective rate is always cgRate relative to the gross profit (spread).
  return cgRate;
}

/**
 * Net per share for ESPP.
 *
 * Sale ≥ purchase-date FMV: discount portion taxed at marginal, appreciation
 * above purchase-date FMV at the capital-gains rate (capital track) or
 * marginal (ordinary track).
 *
 * Sale < purchase-date FMV (sell-below-FMV cap): the discount-as-marginal
 * portion is the actual realized gain max(0, sale − purchase_price), not the
 * theoretical 15% discount. Capital component clamps to zero.
 */
export function esppNetPerShare(
  purchasePrice: number,
  purchaseDateFmv: number,
  salePrice: number,
  marginal: number,
  cgRate: number,
  capitalTrack: boolean,
): number {
  if (salePrice < purchaseDateFmv) {
    const discount = Math.max(0, salePrice - purchasePrice);
    return Math.max(0, salePrice - discount * marginal);
  }
  const discount = Math.max(0, purchaseDateFmv - purchasePrice);
  const appreciation = salePrice - purchaseDateFmv;
  const appreciationRate = capitalTrack ? cgRate : marginal;
  return Math.max(0, salePrice - discount * marginal - appreciation * appreciationRate);
}

export function esppEffectiveTaxRate(
  purchasePrice: number,
  purchaseDateFmv: number,
  salePrice: number,
  marginal: number,
  cgRate: number,
  capitalTrack: boolean,
): number {
  if (salePrice <= 0) return 0;
  const net = esppNetPerShare(purchasePrice, purchaseDateFmv, salePrice, marginal, cgRate, capitalTrack);
  return 1 - net / salePrice;
}

export interface TaxBreakdown {
  title: string;
  lines: string[];
  note?: string;
  /** Marginal portfolio יסף attributable to this lot, if shown in the tooltip. */
  surtaxLine?: string;
}

const fmtILS0 = new Intl.NumberFormat('he-IL', { maximumFractionDigits: 0 });
const fmtILS2 = new Intl.NumberFormat('he-IL', { maximumFractionDigits: 2 });
const fmtUSD = new Intl.NumberFormat('en-US', { minimumFractionDigits: 2, maximumFractionDigits: 2 });

function pct(r: number): string {
  return `${(r * 100).toFixed(0)}%`;
}

export function rsuTaxBreakdown(
  fmvAtGrant: number,
  salePrice: number,
  rate: number,
  cgRate: number,
  capitalTrack: boolean,
  qty: number,
  fx: number,
): TaxBreakdown {
  const lines: string[] = [];
  if (capitalTrack && salePrice >= fmvAtGrant) {
    const empPerShareNIS = fmvAtGrant * rate * fx;
    const apprPerShareNIS = (salePrice - fmvAtGrant) * cgRate * fx;
    const totalEmp = empPerShareNIS * qty;
    const totalAppr = apprPerShareNIS * qty;
    lines.push(`חלק פירותי: $${fmtUSD.format(fmvAtGrant)} × ${pct(rate)} = ${fmtILS2.format(empPerShareNIS)} ₪/מניה`);
    lines.push(`רווח הון: $${fmtUSD.format(salePrice - fmvAtGrant)} × ${pct(cgRate)} = ${fmtILS2.format(apprPerShareNIS)} ₪/מניה`);
    lines.push(`סה"כ × ${fmtILS0.format(qty)} מניות = ${fmtILS0.format(totalEmp + totalAppr)} ₪`);
    return {
      title: 'חישוב מס — RSU מסלול הוני',
      lines,
      note: `חלק ה-FMV-במענק כהכנסה פירותית; העלייה מעל זה כרווח הון ${pct(cgRate)}.`,
    };
  }
  if (capitalTrack && salePrice < fmvAtGrant) {
    const taxPerShareNIS = salePrice * rate * fx;
    lines.push(`כל המכירה במס שולי: $${fmtUSD.format(salePrice)} × ${pct(rate)} = ${fmtILS2.format(taxPerShareNIS)} ₪/מניה`);
    lines.push(`סה"כ × ${fmtILS0.format(qty)} = ${fmtILS0.format(taxPerShareNIS * qty)} ₪`);
    return {
      title: 'חישוב מס — RSU מסלול הוני (price < FMV)',
      lines,
      note: 'מחיר המכירה נמוך מה-FMV במענק → אין רווח הון להכיר; כל הסכום פירותי.',
    };
  }
  const taxPerShareNIS = salePrice * rate * fx;
  lines.push(`כל המכירה במס שולי: $${fmtUSD.format(salePrice)} × ${pct(rate)} = ${fmtILS2.format(taxPerShareNIS)} ₪/מניה`);
  lines.push(`סה"כ × ${fmtILS0.format(qty)} = ${fmtILS0.format(taxPerShareNIS * qty)} ₪`);
  return {
    title: 'חישוב מס — RSU מסלול רגיל',
    lines,
    note: 'פחות מ-24 חודשים מתום שנת המס של ההענקה → אין מסלול הוני, הכל פירותי.',
  };
}

export function optionTaxBreakdown(
  strike: number,
  salePrice: number,
  cgRate: number,
  qty: number,
  fx: number,
): TaxBreakdown {
  const spread = Math.max(0, salePrice - strike);
  const taxPerShareNIS = spread * cgRate * fx;
  return {
    title: 'חישוב מס — אופציות NQ',
    lines: [
      `Spread: $${fmtUSD.format(salePrice)} − $${fmtUSD.format(strike)} = $${fmtUSD.format(spread)}`,
      `מס רווח הון: $${fmtUSD.format(spread)} × ${pct(cgRate)} = ${fmtILS2.format(taxPerShareNIS)} ₪/מניה`,
      `סה"כ × ${fmtILS0.format(qty)} = ${fmtILS0.format(taxPerShareNIS * qty)} ₪`,
    ],
    note: `לפני הנפקה: strike ≈ FMV במענק, ולכן רוב/כל הספרד נחשב רווח הון ${pct(cgRate)}.`,
  };
}

export function esppTaxBreakdown(
  purchasePrice: number,
  purchaseDateFmv: number,
  salePrice: number,
  marginalRateValue: number,
  cgRate: number,
  capitalTrack: boolean,
  qty: number,
  fx: number,
): TaxBreakdown {
  if (salePrice < purchaseDateFmv) {
    const discount = Math.max(0, salePrice - purchasePrice);
    const taxPerShareNIS = discount * marginalRateValue * fx;
    return {
      title: 'חישוב מס — ESPP (price < FMV)',
      lines: [
        `הרווח בפועל: $${fmtUSD.format(salePrice)} − $${fmtUSD.format(purchasePrice)} = $${fmtUSD.format(discount)}`,
        `מס שולי על ההפרש: × ${pct(marginalRateValue)} = ${fmtILS2.format(taxPerShareNIS)} ₪/מניה`,
        `סה"כ × ${fmtILS0.format(qty)} = ${fmtILS0.format(taxPerShareNIS * qty)} ₪`,
      ],
      note: 'מחיר המכירה נמוך מה-FMV ברכישה → ההכנסה הפירותית מוגבלת לרווח בפועל, אין רווח הון.',
    };
  }
  const discount = Math.max(0, purchaseDateFmv - purchasePrice);
  const appreciation = Math.max(0, salePrice - purchaseDateFmv);
  const apprRate = capitalTrack ? cgRate : marginalRateValue;
  const discountTaxNIS = discount * marginalRateValue * fx;
  const apprTaxNIS = appreciation * apprRate * fx;
  const total = (discountTaxNIS + apprTaxNIS) * qty;
  const lines = [
    `הנחה: $${fmtUSD.format(discount)} × ${pct(marginalRateValue)} = ${fmtILS2.format(discountTaxNIS)} ₪/מניה`,
    `עליית ערך: $${fmtUSD.format(appreciation)} × ${pct(apprRate)}${capitalTrack ? ' (הון)' : ' (שולי)'} = ${fmtILS2.format(apprTaxNIS)} ₪/מניה`,
    `סה"כ × ${fmtILS0.format(qty)} = ${fmtILS0.format(total)} ₪`,
  ];
  return {
    title: capitalTrack ? 'חישוב מס — ESPP מסלול הוני' : 'חישוב מס — ESPP מסלול רגיל',
    lines,
    note: 'השיעור האפקטיבי המוצג מחושב יחסית לברוטו המכירה (כולל ה-cost basis). השיעור על הרווח עצמו קרוב לשיעור השולי.',
  };
}

File: .cursor/skills/jfrog-equity-analyzer/tax_calc.py
"""
Israeli equity-compensation tax math, derived from tax-rules.md.

Pure functions, no I/O, no external deps. Every constant and formula here
mirrors the documentation in tax-rules.md verbatim — change one and update
the other, otherwise the sentinel tests in test_tax_calc.py will fail.

All amounts are in NIS unless otherwise noted. Annual figures use the 2026
brackets and ceilings.
"""
from __future__ import annotations

from dataclasses import dataclass
from datetime import date
from typing import Literal, TypedDict

Track = Literal["capital", "ordinary"]


SURTAX_THRESHOLD: float = 721_560.0
SURTAX_GENERAL_RATE: float = 0.03
SURTAX_CAPITAL_EXTRA_RATE: float = 0.02

CAPITAL_GAINS_RATE_BASE: float = 0.25
CAPITAL_GAINS_RATE_HIGH: float = 0.30

ETRADE_TRUSTEE_MARGINAL_WITHHOLD: float = 0.62
ETRADE_TRUSTEE_CAPITAL_WITHHOLD: float = 0.28


INCOME_TAX_BRACKETS_2026: list[tuple[float, float]] = [
    (84_120.0, 0.10),
    (120_720.0, 0.14),
    (228_000.0, 0.20),
    (301_200.0, 0.31),
    (560_280.0, 0.35),
    (721_560.0, 0.47),
    (float("inf"), 0.50),
]


BL_REDUCED_CEILING: float = 7_703.0
BL_MONTHLY_CEILING: float = 51_910.0
BL_REDUCED_RATE: float = 0.004
HEALTH_REDUCED_RATE: float = 0.031
BL_FULL_RATE: float = 0.07
HEALTH_FULL_RATE: float = 0.05
BL_HEALTH_REDUCED_TOTAL: float = BL_REDUCED_RATE + HEALTH_REDUCED_RATE
BL_HEALTH_FULL_TOTAL: float = BL_FULL_RATE + HEALTH_FULL_RATE


def marginal_tax_on(annual_income_nis: float) -> float:
    """Annual income tax (incl. 3% yasaf above 721,560) on a given annual gross.

    Piecewise sum across the 2026 brackets. Returns 0 for non-positive income.
    """
    if annual_income_nis <= 0:
        return 0.0
    tax = 0.0
    prev_top = 0.0
    for top, rate in INCOME_TAX_BRACKETS_2026:
        if annual_income_nis <= top:
            tax += (annual_income_nis - prev_top) * rate
            return tax
        tax += (top - prev_top) * rate
        prev_top = top
    return tax


def capital_gains_rate(
    _annual_income_nis: float,
    controlling_shareholder: bool = False,
) -> float:
    """Effective capital-gains rate for a Section 102 honi sale.

    25% by default, 30% only for controlling shareholders (10%+ holdings ever).
    Portfolio-level מס יסף (3% + 2% since 2025) is computed separately via
    `compute_surtax`.
    """
    if controlling_shareholder:
        return CAPITAL_GAINS_RATE_HIGH
    return CAPITAL_GAINS_RATE_BASE


@dataclass(frozen=True)
class SurtaxResult:
    total_income_nis: float
    capital_income_total_nis: float
    yasaf3_nis: float
    yasaf2_nis: float
    total_nis: float


def compute_surtax(
    salary_nis: float,
    jfrog_sale_total_income_nis: float,
    jfrog_sale_capital_source_nis: float,
    other_capital_income_nis: float = 0.0,
) -> SurtaxResult:
    """Portfolio יסף per section 121ב (הוראת ביצוע 5/2025): 3% on total, 2% on capital."""
    other = other_capital_income_nis
    total_income = salary_nis + jfrog_sale_total_income_nis + other
    capital_total = jfrog_sale_capital_source_nis + other
    yasaf3 = SURTAX_GENERAL_RATE * max(0.0, total_income - SURTAX_THRESHOLD)
    yasaf2 = SURTAX_CAPITAL_EXTRA_RATE * max(0.0, capital_total - SURTAX_THRESHOLD)
    total_yasaf = yasaf3 + yasaf2
    return SurtaxResult(
        total_income_nis=total_income,
        capital_income_total_nis=capital_total,
        yasaf3_nis=yasaf3,
        yasaf2_nis=yasaf2,
        total_nis=total_yasaf,
    )


class LotIncomeBreakdown(TypedDict):
    total_income_nis: float
    capital_income_nis: float


def rsu_lot_income(
    fmv_at_grant: float,
    sale_price: float,
    qty: float,
    fx: float,
    capital_track: bool,
) -> LotIncomeBreakdown:
    del fmv_at_grant
    total_income = sale_price * qty * fx
    cap = total_income if capital_track else 0.0
    return LotIncomeBreakdown(total_income_nis=total_income, capital_income_nis=cap)


def option_lot_income(
    strike: float, sale_price: float, qty: float, fx: float
) -> LotIncomeBreakdown:
    spread = max(0.0, sale_price - strike)
    v = spread * qty * fx
    return LotIncomeBreakdown(total_income_nis=v, capital_income_nis=v)


def espp_lot_income(
    purchase_date_fmv: float,
    sale_price: float,
    qty: float,
    fx: float,
    capital_track: bool,
) -> LotIncomeBreakdown:
    del purchase_date_fmv
    total_income = sale_price * qty * fx
    cap = total_income if capital_track else 0.0
    return LotIncomeBreakdown(total_income_nis=total_income, capital_income_nis=cap)


def holding_track(grant_date: date, sale_date: date) -> Track:
    """24-month rule for 102 trustee tracks.

    The clock starts at the END of the tax year of grant. A grant on
    2024-03-15 completes the 24-month period on 2026-12-31, not 2026-03-15.
    Sales on or after that date qualify for the capital-gains track.
    """
    end_of_grant_year = date(grant_date.year, 12, 31)
    qualifying_year = end_of_grant_year.year + 2
    qualifying_date = date(qualifying_year, 12, 31)
    if sale_date >= qualifying_date:
        return "capital"
    return "ordinary"


def bituach_leumi_employee(monthly_income_nis: float) -> float:
    """Combined BL + health tax on a single month's income (employee share).

    Reduced rates apply to the first 7,703 NIS, full rates between 7,703 and
    the monthly ceiling of 51,910 NIS, no contribution above the ceiling.
    """
    if monthly_income_nis <= 0:
        return 0.0
    reduced = min(monthly_income_nis, BL_REDUCED_CEILING)
    above_reduced = max(0.0, min(monthly_income_nis, BL_MONTHLY_CEILING) - BL_REDUCED_CEILING)
    return reduced * BL_HEALTH_REDUCED_TOTAL + above_reduced * BL_HEALTH_FULL_TOTAL


def rsu_net_per_share(
    price: float,
    fmv_grant: float,
    marginal: float,
    cg_rate: float,
    track: Track = "capital",
) -> float:
    """Net per share for one RSU sale.

    On the ordinary track (or any 102 path before the 24-month clock has
    elapsed) the entire sale price is taxed at the marginal rate.

    On the capital track the gain is split between FMV-at-grant (taxed at
    marginal) and the appreciation above it (taxed at cg_rate). When the
    sale price is at or below FMV-at-grant the rule collapses: all proceeds
    become ordinary income, and the capital component clamps to zero — the
    sell-below-FMV cap.
    """
    if track == "ordinary":
        return price * (1 - marginal)
    if price >= fmv_grant:
        return price - fmv_grant * marginal - (price - fmv_grant) * cg_rate
    return price * (1 - marginal)


def options_net_per_share(price: float, strike: float, cg_rate: float) -> float:
    """Net per share for one NQ option sale under 102 honi.

    Strike ≈ FMV at grant for pre-IPO options, so the entire spread is
    capital gain. Underwater options return zero.
    """
    return max(0.0, price - strike) * (1 - cg_rate)


def espp_net_per_share(
    price: float,
    purchase_price: float,
    purchase_fmv: float,
    marginal: float,
    cg_rate: float,
) -> float:
    """Net per share for one ESPP sale.

    Discount portion (purchase_fmv − purchase_price) is taxed at marginal,
    appreciation above purchase_fmv is taxed at cg_rate.

    Sell-below-FMV cap: when price < purchase_fmv the marginal-as-discount
    portion is the actual realized gain max(0, price − purchase_price)
    rather than the theoretical discount. Capital gain clamps to zero.
    """
    if price >= purchase_fmv:
        discount = max(0.0, purchase_fmv - purchase_price)
        appreciation = price - purchase_fmv
        return price - discount * marginal - appreciation * cg_rate
    discount = max(0.0, price - purchase_price)
    return price - discount * marginal
